src/utils/provider.ts

```diff
@@ -0,0 +1,37 @@
+import { useContext, useState } from 'react'
+import { defaultConfig } from './config'
+import { wrapCache } from './cache'
+import { isFunction, UNDEFINED } from './helper'
+import { SWRConfigContext } from './config-context'
+import { SWRGlobalState, GlobalState } from './global-state'
+
+import { Cache, ProviderOptions } from '../types'
+
+export const useSWRProvider = <Data = any>(
+  initialize?:
+    | ((currentCache: Readonly<Cache>) => Cache)
+    | Partial<ProviderOptions>,
+  options?: Partial<ProviderOptions>
+) => {
+  const cache = useContext(SWRConfigContext).cache || defaultConfig.cache
+
+  // We use a state to keep the cache instance and make sure it's only
+  // initialized once.
+  const [newProvider] = useState(() => {
```

Comment:
This function has two responsibilities:
* create and return a scoped cache
* return a inherited cache

Consider separating them?

the creating cache part can be moved to `SWRConfig` and only keep `useSWRProvider()` for consuming purpose.